feat: Added pokemonDetails page features

NicoEnriquezFFuFm opened the pull request:
- added image_widget_pokemon.dart inside widgets
- added overview_widget.dart inside widgets
- added row_description.dart inside widgets
- added pokemon_details_connector.dart
- added pokemon_details_page.dart
- added pokemon_details_vm.dart
- modified GetPokemonDetailAction
- added additional freezed model files

dart formatted
![image](https://user-images.githubusercontent.com/128012634/228182223-713c8537-94f5-45b4-b77b-6663e679fff8.png)

dart analyzed
![flutter_analyze_detailsPage](https://user-images.githubusercontent.com/128012634/228182342-2f37b249-5a49-47ec-96f3-a714f494880c.PNG)

video:


https://user-images.githubusercontent.com/128012634/228182420-1c05334d-a266-4852-b646-92ca4ea1ff83.mp4

NicoEnriquezFFuFm committed:
feat: Added pokemonDetails page features
- added image_widget_pokemon.dart inside widgets
- added overview_widget.dart inside widgets
- added row_description.dart inside widgets
- added pokemon_details_connector.dart
- added pokemon_details_page.dart
- added pokemon_details_vm.dart
- modified GetPokemonDetailAction
- added additional freezed model files

shawn-ffufm commented on lib/api/model/pokemon_detail.dart:
```diff
@@ -15,6 +17,8 @@ class PokemonDetail with _$PokemonDetail {
     @JsonKey(name: "abilities") List<Abilities>? abilities,
```
haven't i explained before to avoid nullable collection, instead make non-nullable but make sure to always pass an empty collection? 🙂

shawn-ffufm commented on lib/api/model/pokemon_sprites.dart:
```diff
@@ -0,0 +1,16 @@
+@freezed
+class PokemonSprites with _$PokemonSprites {
+  factory PokemonSprites({
+    @JsonKey(name: "back_default") String? backDefault,
```
are these properties expected to become null or no that's why they're nullable?

shawn-ffufm commented on lib/api/model/pokemon_type.dart:
```diff
@@ -0,0 +1,14 @@
+@freezed
+class PokemonType with _$PokemonType {
+  factory PokemonType({
+    @JsonKey(name: "name") String? name,
```
same here, see related feedback above 🙂

shawn-ffufm commented on lib/api/model/pokemon_types.dart:
```diff
@@ -0,0 +1,15 @@
+@freezed
+class PokemonTypes with _$PokemonTypes {
+  factory PokemonTypes({
+    @JsonKey(name: "slot") int? slot,
```
same here, see related feedback above 🙂

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,98 @@
+import 'package:pokedexbootcamp/widgets/image_widget_pokemon_details.dart';
+import 'package:pokedexbootcamp/widgets/row_description.dart';
+
+class PokemonDetailsPage extends StatefulWidget {
```
where's the need why u needed to make this widget stateful? 🙂

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,98 @@
+    required this.pokemonName,
+  }) : super(key: key);
+
+  final Async<PokemonDetail?> pokemonDetail;
```
why is the PokemonDetail nullable? 🙂

shawn-ffufm replied:
any update here? 🙂

NicoEnriquezFFuFm replied:
oh thats strange, I forgot to add the updated one to the commit will make it to the next push

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,98 @@
+
+class _PokemonDetailsPageState extends State<PokemonDetailsPage> {
+  @override
+  void dispose() {
```
why override this function when u don't even use it? 🙂

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,98 @@
+      appBar: AppBar(
+        leading: IconButton(
+          icon: const Icon(Icons.arrow_back, color: Colors.white),
+          onPressed: () => Navigator.of(context).pop(),
```
can u put this in a separate private function to make more readable 🙂

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,98 @@
+        title: const Text(
+          pokemonAppBarTitle,
+          style: TextStyle(
+            fontSize: fontSizeAppBar,
```
no need for extra trailing comma here 🙂

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,98 @@
+        error: (errorMessage) {
+          WidgetsBinding.instance.addPostFrameCallback((_) => _showErrorMessageSnackbar(context, errorMessage));
+          return const Center(
+            child: Text(pokemonOfflineMessageLabel),
```
no need for extra trailing comma here 🙂

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,98 @@
+                mainAxisAlignment: MainAxisAlignment.center,
+                crossAxisAlignment: CrossAxisAlignment.center,
+                children: [
+                  PokemonDetailsImages(pokeImage: data!.sprite!.frontDefault!),
```
why create data property nullable when in the end you're just gonna end of enforcing them to become non-null 🤔

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_vm.dart:
```diff
@@ -0,0 +1,27 @@
+
+class PokemonDetailsVmFactory extends VmFactory<AppState, PokemonDetailsConnector> {
+  @override
+  Vm? fromStore() => PokemonDetailsVm(pokemonDetail: _pokemonDetail());
```
why is Vm nullable?

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_vm.dart:
```diff
@@ -0,0 +1,27 @@
+  @override
+  Vm? fromStore() => PokemonDetailsVm(pokemonDetail: _pokemonDetail());
+
+  Async<PokemonDetail?> _pokemonDetail() {
```
why is PokemonDetail nullable? 🙂

NicoEnriquezFFuFm replied:
My bad, forgot to switch it back

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_vm.dart:
```diff
@@ -0,0 +1,27 @@
+    required this.pokemonDetail,
+  }) : super(equals: [pokemonDetail]);
+
+  final Async<PokemonDetail?> pokemonDetail;
```
see related feedback above 🙂

shawn-ffufm commented on lib/widgets/image_widget_pokemon_details.dart:
```diff
@@ -0,0 +1,31 @@
+    this.pokeImage,
+  }) : super(key: key);
+
+  final dynamic pokeImage;
```
why dynamic?

NicoEnriquezFFuFm replied:
The dynamic variable is for a way to pass the sprite images to the image network.

shawn-ffufm commented on lib/widgets/overview_widget.dart:
```diff
@@ -1,9 +1,13 @@
-  const PokemonOverview({Key? key, required this.pokemonName}) : super(key: key);
+  const PokemonOverview({
+    Key? key,
+    required this.pokemonName,
```
make it as a habit to place required properties to be at the top of everything 🙂

shawn-ffufm commented on lib/widgets/overview_widget.dart:
```diff
@@ -15,27 +19,34 @@ class PokemonOverview extends StatelessWidget {
         GestureDetector(
-          //TODO: onTap will navigate to the details of the pokemon
-          onTap: null,
+          onTap: () => Navigator.push(
```
kindly move this into a separate private function with better naming 🙂

shawn-ffufm commented on lib/widgets/overview_widget.dart:
```diff
@@ -15,27 +19,34 @@ class PokemonOverview extends StatelessWidget {
+            context,
+            MaterialPageRoute(
+              builder: (context) => PokemonDetailsConnector(
+                pokemonName: pokemonName.name,
```
no need for extra trailing comma here 🙂

shawn-ffufm commented on lib/widgets/overview_widget.dart:
```diff
@@ -15,27 +19,34 @@ class PokemonOverview extends StatelessWidget {
                   pokemonName.name,
-                  style: const TextStyle(fontSize: fontSize),
+                  style: const TextStyle(
+                    fontSize: fontSize,
```
no need for extra trailing comma here 🙂

shawn-ffufm commented on lib/widgets/row_description.dart:
```diff
@@ -0,0 +1,35 @@
+        Row(
+          children: [
+            Text(
+              "$title:",
```
what is this?

shawn-ffufm commented on lib/widgets/row_description.dart:
```diff
@@ -0,0 +1,35 @@
+          children: [
+            Text(
+              "$title:",
+              style: const TextStyle(fontSize: fontSize, fontWeight: FontWeight.w600),
```
add extra trailing comma here 🙂

shawn-ffufm reviewed (changes requested):
see pr feedback 🙂

NicoEnriquezFFuFm committed:
feat: Added pokemonDetails page features
(pr feedback)
- added functions for navigation from pop to push
- modified  freezed models to non nullable
- removed some comas in the widget pages and pokemon_details_page
- removed the nullable in PokemonDetails inside pokemon_details_vm
- modifed title variable to titleLabel inside row_description
- modified PokemonDetailsPage from stateful to stateless widget
- modified pokeImage from dynamic to var inside PokemonDetailsImages

NicoEnriquezFFuFm committed:
feat: Added pokemonDetails page features
(pr feedback)
- modified image_widget_pokemon_details.dart from var to String

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_connector.dart:
```diff
@@ -0,0 +1,27 @@
+  Widget build(BuildContext context) {
+    return StoreConnector<AppState, PokemonDetailsVm>(
+      vm: () => PokemonDetailsVmFactory(),
+      onInit: (store) => store.dispatchAsync(GetPokemonDetailAction(pokemonName: pokemonName)),
```
where's the action to clear the pokemon details state during onDispose stage? 🙂

shawn-ffufm reviewed (changes requested):
see minor pr feedback

NicoEnriquezFFuFm committed:
feat: Added pokemonDetails page features
(pr feedback)
- modified PokemonDetailsPage variable constructor from Async<PokemonDetail?> to Async<PokemonDetail>
- added dispose action for pokemondetailspage
- modified pokemon_details_connector, added onDispose inside StoreConnector
